compose/config/interpolation.py

```diff
@@ -26,9 +29,46 @@ def process_item(name, config_dict):
     )
 
 
-def interpolate_value(name, config_key, value, section, mapping):
+def _cast_interpolated(interpolated_value, expected_type):
+
+    if expected_type == "string":
+        return None
+    elif expected_type == "number":
+        try:
+            return int(interpolated_value)
+        except ValueError:
+            return None
+    elif expected_type == "boolean":
+        if "true" == interpolated_value.lower():
+            return True
+        if "false" == interpolated_value.lower():
+            return False
+        return None
+    elif expected_type == "array":
+        if isinstance(interpolated_value, list):
+            return interpolated_value
+        return json.loads("{{\"data\":{0}}}".format(interpolated_value.replace("'", "\"")))["data"]
+
+
+def interpolate_value(name, config_key, value, section, mapping, schema):
     try:
-        return recursive_interpolate(value, mapping)
+        interpolated = recursive_interpolate(value, mapping)
+        if (interpolated != value):
+            # cast as needed
+            if config_key in schema["definitions"]["service"]["properties"]:
+                field_def = schema["definitions"]["service"]["properties"][config_key]
+                if "type" in field_def:
+                    allowed_types = field_def["type"]
+                    if isinstance(allowed_types, list):
+                        for allowed_type in allowed_types:
+                            converted = _cast_interpolated(interpolated, allowed_type)
+                            if converted is not None:
+                                return converted
+                    else:
+                        converted = _cast_interpolated(interpolated, allowed_types)
+                        if converted is not None:
+                            return converted
+        return interpolated
```

Comment:
I think we could avoid doing this type inspection here. In `recursive_interpolate()` we can pull out the part of the schema relevant for the section we're going into and pass that as the value for `schema` when calling itself.

That way `cast_interpolated()` can be called from just the `else` block in `recursive_interpolate()`.